### PR TITLE
remove KryoSerializable from UnsafeRow and UnsafeIndexedSeq

### PR DESCRIPTION
**Pull request by catoverdrive**
This basically involved two main things:

- converting the rangeBounds on OrderedRVDPartitioner from an UnsafeIndexedSeq to an IndexedSeq[Interval], which had the nice side effect of getting rid of the gross-looking `rangeBounds(i).asInstanceOf[Interval]` stuff everywhere. Since we're pulling everything into Annotation-land when we do comparisons against this, we were basically using rangeBounds in this way anyways.

- implementing Annotation.copy as a fully safe copy that gets rid of all instances of UnsafeRow and UnsafeIndexedSeq. This is used for e.g. making the rangeBounds fully safe, and also for when we need to serialize an annotation that could potentially contain UnsafeRows. 

I originally had implemented this as a separate method from the existing copy, but I think the only reason we call Annotation.copy() anyways is so that we can call collect() or otherwise serialize things, so I just removed the other version in favor of this.

I implemented a way to manually serialize UnsafeRow and UnsafeIndexedSeq as byte arrays, but I think  there's basically no reason we should be manually serializing UnsafeRows instead of just serializing the underlying region, so I took it back out.

### CODE REVIEW

File: src/main/scala/is/hail/annotations/Annotation.scala [hunk 1]
```diff
@@ -161,18 +162,29 @@ object Annotation {
...
+          .sorted(t.elementType.ordering.toOrdering).toFastIndexedSeq
+
+        s.map(Annotation.copy(t.elementType, _))
+
+      case t: TDict =>
+        val m = a.asInstanceOf[Map[Annotation, Annotation]]
+          .map { case (k, v) => Row(k, v) }
+          .toArray
+          .sorted(t.elementType.ordering.toOrdering)
+        m.map(Annotation.copy(t.elementType, _))
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/Annotation.scala]: Same.

File: src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala [hunk 2]
```diff
@@ -101,38 +93,40 @@ class OrderedRVDPartitioner(
...
+        case IndexedSeq(x) => IndexedSeq(x.copy(newStart, newEnd, true, true))
+        case IndexedSeq(x1, x2) =>
+          IndexedSeq(x1.copy(start = newStart, includesStart = true),
+            x2.copy(end = newEnd, includesEnd = true))
+        case _ =>
+          rangeBounds.head.copy(start = newStart, includesStart = true)  +:
+            rangeBounds.tail.init :+
+            rangeBounds.last.copy(end = newEnd, includesEnd = true)
+      }
+    copy(rangeBounds = Annotation.copy(rangeBoundsType, newRangeBounds).asInstanceOf[IndexedSeq[Interval]])
```
Comment (cseed) [inline src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala]: The copy here seems excessive, only the input `newRange` Interval can contain region values.

File: src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala [hunk 3]
```diff
@@ -101,38 +93,40 @@ class OrderedRVDPartitioner(
...
-    newRangeBounds(0) = newRangeBounds(0).asInstanceOf[Interval]
-      .copy(start = newStart, includesStart = true)
-    newRangeBounds(newRangeBounds.length - 1) = newRangeBounds(newRangeBounds.length - 1)
-      .asInstanceOf[Interval].copy(end = newEnd, includesEnd = true)
-    copy(rangeBounds = UnsafeIndexedSeq(rangeBoundsType, newRangeBounds))
+    val newStart = pkType.ordering.min(range.start, Annotation.copy(pkType, newRange.start))
+    val newEnd = pkType.ordering.max(range.end, Annotation.copy(pkType, newRange.end))
+    val newRangeBounds =
+      rangeBounds match {
+        case IndexedSeq(x) => IndexedSeq(x.copy(newStart, newEnd, true, true))
```
Comment (cseed) [inline src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala]: Hmm, don't call copy if you're going to change ever field.

File: src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala [hunk 4]
```diff
@@ -101,38 +93,40 @@ class OrderedRVDPartitioner(
...
+    val newRangeBounds =
+      rangeBounds match {
+        case IndexedSeq(x) => IndexedSeq(Interval(newStart, newEnd, true, true))
+        case IndexedSeq(x1, x2) =>
+          IndexedSeq(x1.copy(start = newStart, includesStart = true),
+            x2.copy(end = newEnd, includesEnd = true))
+        case _ =>
+          rangeBounds.head.copy(start = newStart, includesStart = true)  +:
+            rangeBounds.tail.init :+
+            rangeBounds.last.copy(end = newEnd, includesEnd = true)
```
Comment (danking) [inline src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala]: This is much easier to read with the match statement! I like it.

For the last case you can also do:

```scala
case head +: tail => head.copy(...) +: tail.init :+ tail.last.copy(...)
```

**Pull request merged**